### PR TITLE
JP-3668: Move tweakreg into stcal

### LINKED ISSUE #8594: tweakreg to stcal
_Issue [JP-3668](https://jira.stsci.edu/browse/JP-3668) was created on JIRA by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter):_

The portion of the JWST tweakreg step that does the actual tweakreg part (as opposed to source finding) performs an identical task to its counterpart in Roman.  Because of that, it's sensible to move this piece of code into stcal to avoid duplicating code and to ensure any bugfixes or improvements benefit both telescopes.  As a recent example, some memory optimizations to the step have been made by [Brett Graham](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bgraham) that exist only on the JWST side at present.

The [stcal ticket](https://jira.stsci.edu/browse/AL-850) covers the work in that repository.  On the JWST side, the stcal version of tweakreg will simply need to be called within the existing framework of the tweakreg step.

### PR DESCRIPTION
**Pull request by emolter**
Resolves [JP-3668](https://jira.stsci.edu/browse/JP-3668)

Closes #8594

The portions of the tweakreg step that are shared between jwst and romancal are being moved into the stcal repository; see [this stcal PR](https://github.com/spacetelescope/stcal/pull/267).  This PR updates the jwst pipeline accordingly.

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by emolter** [comment]
one-liner change to make the version of `update_s_region_imaging` in stcal identical to what used to be in the jwst repo did indeed [fix the regtest failures](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1599/).  Unfortunately, these seem unrelated to the single unit test failure `tweakreg/tests/test_multichip_jwst.py:test_multichip_alignment_step`.  Still tracking that one down.

File: jwst/tweakreg/tests/test_multichip_jwst.py
Comment (mcara) [inline jwst/tweakreg/tests/test_multichip_jwst.py:228]:
```suggestion
    monkeypatch.setattr(tweakreg_step.twk, 'align_wcs', _align_wcs)
```

File: jwst/tweakreg/tests/test_multichip_jwst.py
Comment (mcara) [inline jwst/tweakreg/tests/test_multichip_jwst.py:444]:
```suggestion
    monkeypatch.setattr(tweakreg_step.twk, 'align_wcs', _align_wcs)
```

File: jwst/tweakreg/tests/test_multichip_jwst.py
Comment (mcara) [inline jwst/tweakreg/tests/test_multichip_jwst.py:314]:
```suggestion
    monkeypatch.setattr(tweakreg_step.twk, 'align_wcs', _align_wcs)
```

**Review by mcara — CHANGES_REQUESTED** [review]
Failing tests are due to monkeypatch not patching align_wcs.

File: jwst/tweakreg/tests/test_multichip_jwst.py
Comment (emolter) [reply jwst/tweakreg/tests/test_multichip_jwst.py]: Thanks @mcara, sure enough, it works for me locally

**Review by mcara — COMMENTED** [review]
I will withhold my final review pending review of the sister PR in stcal. Both PRs look good but I would like to look into the issue of data model interface. At this time this PR is "Approved" but I do not want it merged yet.

**Review by mcara — APPROVED** [review]
LGTM

**Comment by emolter** [comment]
started new [regression test run here](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/lastBuild/)

**Comment by emolter** [comment]
new regression tests [started here](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/lastBuild/), after minor changes to both stcal and jwst sides

**Comment by emolter** [comment]
Another new [set of regtests here](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1618/)

**Comment by emolter** [comment]
ok, hopefully the last round - I realized I had inadvertently changed the default value of the `center` parameter in the `update_s_region_imaging()` function that has now been renamed to `compute_s_region_imaging()` and moved to stcal.  [Regression tests started here](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1619/).  The offenders from the last run were passing locally, at least.

**Comment by emolter** [comment]
[regression test run](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1621/) covering (attempted) fix of the final few failing S_REGION issues

**Review by melanieclarke — COMMENTED** [review]
I'll defer to all the expert reviews already done for this change. From a cursory review, this looks ready to go to me.

It looks like pyproject.toml needs an update after the stcal change is merged, so I'll hold off on approval, pending that coordination.

**Review by nden — APPROVED** [review]
LGTM

**Comment by emolter** [comment]
> I'll defer to all the expert reviews already done for this change. From a cursory review, this looks ready to go to me.
> 
> It looks like pyproject.toml needs an update after the stcal change is merged, so I'll hold off on approval, pending that coordination.

I just merged the stcal side since Nadia gave the approval.  The correct thing to pin is stcal@main, right?  And wait for an stcal release to change it to 1.8.0 later? @melanieclarke

**Comment by melanieclarke** [comment]
> I just merged the stcal side since Nadia gave the approval. The correct thing to pin is stcal@main, right? And wait for an stcal release to change it to 1.8.0 later? @melanieclarke

Sounds right to me. @braingram?

File: pyproject.toml
Comment (braingram) [inline pyproject.toml]:
```suggestion
    "stcal@git+https://github.com/spacetelescope/stcal@main",
```

Comment (emolter) [reply pyproject.toml]: whoops

**Review by melanieclarke — APPROVED** [review]
Okay, all looks good now!

**Pull request merged**
